File: wire/delta.go
package wire

import (
	"encoding/binary"
	"errors"
)

// The v2 frame-delta container (normative, ABI.md §4.5). Every send/identical
// carries this variable-length, little-endian, index-addressed run-list:
//
//	Header (9 bytes):
//	  u8  flags    bit0 = keyframe (1 = full-frame keyframe); all other bits MUST be zero
//	  u32 epoch    the host-issued epoch this delta is computed against (0 on a fresh instance)
//	  u16 runCount number of runs (keyframe: exactly 1; no-change: 0)
//	  u8  rows     grid geometry; MUST be 24 in v2
//	  u8  cols     grid geometry; MUST be 80 in v2
//	then runCount runs, each:
//	  u16 startIndex  first cell index (0..FrameCells-1, == row*Cols+col)
//	  u16 runLen      1..FrameCells, consecutive changed cells
//	  runLen × CellBytes  packed canonical-zero cells (PutCell output)
//
// These encoders/validators are index-addressed (mirroring PutCell/GetCell),
// never the appending Buf, so they are allocation-free over caller-owned
// scratch — the SDK's leaking-GC steady-state requirement.
const (
	// DeltaHeaderBytes is the fixed container header length.
	DeltaHeaderBytes = 9
	// RunHeaderBytes is the per-run prefix (u16 startIndex + u16 runLen).
	RunHeaderBytes = 4
	// KeyframeBytes is the keyframe form's exact size and the worst-case bound:
	// 9-byte header (bit0 set) + one run {start=0, len=FrameCells} + the full
	// 46080-byte packed grid = 46093. The SDK budget rule ships the keyframe
	// when an encoded delta would meet or exceed this (inclusive).
	KeyframeBytes = DeltaHeaderBytes + RunHeaderBytes + FrameBytes // 46093

	// FlagKeyframe is header flags bit0: the payload is a self-contained keyframe.
	FlagKeyframe = 0x01
	// flagsKnown is the set of assigned flag bits; any bit outside it is rejected.
	flagsKnown = FlagKeyframe
)

// MaxDeltaBytes is a scratch-buffer cap big enough for any delta this package
// emits: the run-list worst case never exceeds the keyframe form (the SDK caps
// it there), so KeyframeBytes is a sufficient and exact ceiling.
const MaxDeltaBytes = KeyframeBytes

// errMalformedDelta is the single non-fatal error every malformed-container
// path returns; the host logs "dropped malformed delta", drops, and bumps the
// slot epoch. Validators NEVER panic and NEVER read out of bounds.
var errMalformedDelta = errors.New("wire: malformed frame delta")

// cellEqual reports whether the 24-byte cell at byte offset o is identical in a
// and b. Under the canonical-zero rule cell equality IS a 24-byte memcmp, done
// here as three uint64 loads (the cell is 24 bytes; o is a multiple of 24).
func cellEqual(a, b []byte, o int) bool {
	return binary.LittleEndian.Uint64(a[o:]) == binary.LittleEndian.Uint64(b[o:]) &&
		binary.LittleEndian.Uint64(a[o+8:]) == binary.LittleEndian.Uint64(b[o+8:]) &&
		binary.LittleEndian.Uint64(a[o+16:]) == binary.LittleEndian.Uint64(b[o+16:])
}

// putDeltaHeader writes the 9-byte container header into dst[0:9].
func putDeltaHeader(dst []byte, keyframe bool, epoch uint32, runCount int) {
	if keyframe {
		dst[0] = FlagKeyframe
	} else {
		dst[0] = 0
	}
	binary.LittleEndian.PutUint32(dst[1:], epoch)            // u32 epoch
	binary.LittleEndian.PutUint16(dst[5:], uint16(runCount)) // u16 runCount
	dst[7] = Rows                                            // u8 rows  (24)
	dst[8] = Cols                                            // u8 cols  (80)
}

// BuildFrameDelta is the reference delta encoder: it diffs the packed frames
// base vs next into the caller-provided scratch dst (sized to at least
// MaxDeltaBytes) and returns the number of payload bytes. It coalesces changed
// cells into MAXIMAL runs of consecutive changed cells, greedy left-to-right
// (gap = 0): a single unchanged cell between two changed spans forces two runs.
// That determinism is what makes cross-implementation golden vectors
// byte-identical. epoch is stamped into the header (the slot's host-issued
// epoch). It allocates nothing.
//
// A runCount==0 result (the 9-byte header alone) is the canonical no-change
// delta. The keyframe form is NOT produced here — callers apply the budget rule
// (>= KeyframeBytes ⇒ BuildKeyframe) on the returned length.
func BuildFrameDelta(base, next, dst []byte, epoch uint32) int {
	p := DeltaHeaderBytes
	runs := 0
	i := 0
	for i < FrameCells {
		if cellEqual(base, next, i*CellBytes) {
			i++
			continue
		}
		start := i
		for i < FrameCells && !cellEqual(base, next, i*CellBytes) {
			i++
		}
		runLen := i - start
		binary.LittleEndian.PutUint16(dst[p:], uint16(start))
		p += 2
		binary.LittleEndian.PutUint16(dst[p:], uint16(runLen))
		p += 2
		copy(dst[p:], next[start*CellBytes:(start+runLen)*CellBytes])
		p += runLen * CellBytes
		runs++
	}
	putDeltaHeader(dst, false, epoch, runs)
	return p
}

// BuildKeyframe writes the keyframe form into dst: a 9-byte header (flags bit0
// set), one run {startIndex=0, runLen=FrameCells} and the full packed grid
// next. Returns KeyframeBytes (46093). It allocates nothing.
func BuildKeyframe(next, dst []byte, epoch uint32) int {
	putDeltaHeader(dst, true, epoch, 1)
	p := DeltaHeaderBytes
	binary.LittleEndian.PutUint16(dst[p:], 0)            // start index 0
	binary.LittleEndian.PutUint16(dst[p+2:], FrameCells) // run length 1920
	p += RunHeaderBytes
	p += copy(dst[p:], next)
	return p
}

// DeltaEpoch reads the epoch field from a container header without validating
// the rest. Callers use it to decide acceptance; it returns 0 on a short read.
func DeltaEpoch(b []byte) uint32 {
	if len(b) < DeltaHeaderBytes {
		return 0
	}
	return binary.LittleEndian.Uint32(b[1:])
}

// IsKeyframe reports whether the container's flags carry the keyframe bit. It
// returns false on a short read.
func IsKeyframe(b []byte) bool {
	return len(b) >= 1 && b[0]&FlagKeyframe != 0
}

// CheckFrameDelta validates a frame-delta container structurally without ever
// panicking or reading out of bounds (the drop-not-fatal contract). It checks:
//   - len >= 9 (header present)
//   - geometry bytes == (Rows, Cols) == (24, 80)
//   - no unknown flag bits set (only bit0 is assigned in v2 — rejected, not ignored)
//   - runCount consistent with body length: 9 + Σ(4 + runLen*24) == len(b)
//   - every run in-bounds: startIndex+runLen <= FrameCells
//   - runs strictly ascending and non-overlapping: start[i] >= start[i-1]+len[i-1]
//
// A short read degrades to errMalformedDelta. It does NOT require runs to be
// minimal, greedy, or true diffs — host acceptance is the envelope, so a
// hand-rolled guest's structurally valid container passes.
func CheckFrameDelta(b []byte) error {
	if len(b) < DeltaHeaderBytes {
		return errMalformedDelta
	}
	if b[0]&^flagsKnown != 0 {
		return errMalformedDelta // unknown flag bit set
	}
	if b[7] != Rows || b[8] != Cols {
		return errMalformedDelta // wrong geometry
	}
	runCount := int(binary.LittleEndian.Uint16(b[5:]))
	p := DeltaHeaderBytes
	prevEnd := 0 // first allowed startIndex
	for k := 0; k < runCount; k++ {
		if p+RunHeaderBytes > len(b) {
			return errMalformedDelta
		}
		start := int(binary.LittleEndian.Uint16(b[p:]))
		runLen := int(binary.LittleEndian.Uint16(b[p+2:]))
		if runLen == 0 || start+runLen > FrameCells {
			return errMalformedDelta // empty or out-of-bounds run
		}
		if start < prevEnd {
			return errMalformedDelta // not strictly ascending / overlapping
		}
		p += RunHeaderBytes
		body := runLen * CellBytes
		if p+body > len(b) {
			return errMalformedDelta
		}
		p += body
		prevEnd = start + runLen
	}
	if p != len(b) {
		return errMalformedDelta // trailing bytes / length disagrees with runCount
	}
	return nil
}

// ApplyFrameDelta applies a (validated) delta in place to prev, a FrameBytes
// baseline buffer: each run's cells are copied into prev at startIndex*24, so a
// keyframe (one full-cover run) overwrites all 1920 cells. It re-validates
// structurally (so a caller may apply untrusted bytes safely) and returns
// errMalformedDelta on any malformation, never panicking, never reading OOB,
// and never partially mutating prev on a malformed container (it validates
// fully first). It allocates nothing.
func ApplyFrameDelta(prev, delta []byte) error {
	if len(prev) != FrameBytes {
		return errMalformedDelta
	}
	if err := CheckFrameDelta(delta); err != nil {
		return err
	}
	runCount := int(binary.LittleEndian.Uint16(delta[5:]))
	p := DeltaHeaderBytes
	for k := 0; k < runCount; k++ {
		start := int(binary.LittleEndian.Uint16(delta[p:]))
		runLen := int(binary.LittleEndian.Uint16(delta[p+2:]))
		p += RunHeaderBytes
		body := runLen * CellBytes
		copy(prev[start*CellBytes:start*CellBytes+body], delta[p:p+body])
		p += body
	}
	return nil
}

File: rust/src/delta.rs
//! The ABI v2 frame-delta container (ABI.md §4.5) — the frame payload of every
//! `send`/`identical`. This is the byte-verified RUN-LIST encoder, moved here
//! from `kit/crossverify` (which remains its golden-vector verification
//! harness, path-depending on this crate): maximal-greedy runs, inclusive
//! keyframe threshold, always-present fixed-width epoch, canonical-zero cells.
//!
//! 9-byte header: `u8 flags` (bit0 = keyframe), `u32 epoch` (host-issued; the
//! guest stamps the epoch the host last returned for the slot, 0 on a fresh
//! instance), `u16 runCount`, `u8 rows = 24`, `u8 cols = 80`. Then `runCount`
//! runs of `{u16 startIndex, u16 runLen, runLen * 24 packed cells}`, each a
//! maximal span of CONSECUTIVE changed cells (greedy left-to-right, gap = 0).
//!
//! Not author-facing: `#[doc(hidden)]` public only so the crossverify harness
//! can bind this exact encoder byte-identical to the Go reference.

use crate::frame::{CELL_BYTES, FRAME_BYTES, FRAME_CELLS};

/// 9-byte container header.
pub const DELTA_HEADER_BYTES: usize = 9;
/// Per-run prefix: u16 startIndex + u16 runLen.
pub const RUN_HEADER_BYTES: usize = 4;
/// Header flags bit0: the payload is a self-contained keyframe (full frame).
pub const FLAG_KEYFRAME: u8 = 0x01;
/// Keyframe form / worst-case size: header + one run over all 1920 cells + grid.
pub const KEYFRAME_BYTES: usize = DELTA_HEADER_BYTES + RUN_HEADER_BYTES + FRAME_BYTES; // 46093

/// Grid geometry carried in the last two header bytes (host validates `(24,80)`).
pub const GEOMETRY_ROWS: u8 = crate::frame::ROWS as u8;
pub const GEOMETRY_COLS: u8 = crate::frame::COLS as u8;

/// Write the normative 9-byte header into `dst[0..9]`.
#[inline]
fn put_header(dst: &mut [u8], keyframe: bool, epoch: u32, run_count: usize) {
    dst[0] = if keyframe { FLAG_KEYFRAME } else { 0 };
    dst[1..5].copy_from_slice(&epoch.to_le_bytes());
    dst[5..7].copy_from_slice(&(run_count as u16).to_le_bytes());
    dst[7] = GEOMETRY_ROWS;
    dst[8] = GEOMETRY_COLS;
}

/// Whether the 24-byte cell at offset `o` is identical in `a` and `b`. Under
/// the canonical-zero rule this IS a 24-byte memcmp.
#[inline]
fn cell_equal(a: &[u8], b: &[u8], o: usize) -> bool {
    a[o..o + CELL_BYTES] == b[o..o + CELL_BYTES]
}

/// Encode the maximal-greedy RUN-LIST delta from `prev` to `next` into `dst`.
/// Returns the payload byte length. A `runCount == 0` payload (the header
/// alone) is the canonical "no change" delta. Allocates nothing.
pub fn encode_run_list(prev: &[u8], next: &[u8], dst: &mut [u8], epoch: u32) -> usize {
    debug_assert!(prev.len() >= FRAME_BYTES && next.len() >= FRAME_BYTES);
    let mut p = DELTA_HEADER_BYTES;
    let mut runs = 0usize;
    let mut i = 0usize;
    while i < FRAME_CELLS {
        if cell_equal(prev, next, i * CELL_BYTES) {
            i += 1;
            continue;
        }
        let start = i;
        while i < FRAME_CELLS && !cell_equal(prev, next, i * CELL_BYTES) {
            i += 1;
        }
        let run_len = i - start;
        dst[p..p + 2].copy_from_slice(&(start as u16).to_le_bytes());
        dst[p + 2..p + 4].copy_from_slice(&(run_len as u16).to_le_bytes());
        p += RUN_HEADER_BYTES;
        let src = start * CELL_BYTES;
        let n = run_len * CELL_BYTES;
        dst[p..p + n].copy_from_slice(&next[src..src + n]);
        p += n;
        runs += 1;
    }
    put_header(dst, false, epoch, runs);
    debug_assert_eq!(p, DELTA_HEADER_BYTES + runs * RUN_HEADER_BYTES + payload_cells(dst, runs) * CELL_BYTES);
    p
}

/// Sum of run lengths in an encoded payload — debug-assert support only.
#[cfg(debug_assertions)]
fn payload_cells(dst: &[u8], runs: usize) -> usize {
    let mut p = DELTA_HEADER_BYTES;
    let mut total = 0usize;
    for _ in 0..runs {
        let len = u16::from_le_bytes([dst[p + 2], dst[p + 3]]) as usize;
        total += len;
        p += RUN_HEADER_BYTES + len * CELL_BYTES;
    }
    total
}
#[cfg(not(debug_assertions))]
fn payload_cells(_dst: &[u8], _runs: usize) -> usize {
    0
}

/// Emit the KEYFRAME form: header (flags bit0) + one run `{start=0, len=1920}`
/// + the full 46080-byte packed grid (`KEYFRAME_BYTES`). The bootstrap /
/// full-frame / worst-case payload.
pub fn encode_keyframe(next: &[u8], dst: &mut [u8], epoch: u32) -> usize {
    put_header(dst, true, epoch, 1);
    let mut p = DELTA_HEADER_BYTES;
    dst[p..p + 2].copy_from_slice(&0u16.to_le_bytes()); // start 0
    dst[p + 2..p + 4].copy_from_slice(&(FRAME_CELLS as u16).to_le_bytes()); // len 1920
    p += RUN_HEADER_BYTES;
    dst[p..p + FRAME_BYTES].copy_from_slice(&next[..FRAME_BYTES]);
    p + FRAME_BYTES
}

/// Encode the delta, falling back to the keyframe form when the run-list would
/// meet or exceed the keyframe size (inclusive `>=`), bounding the worst case
/// at exactly `KEYFRAME_BYTES`. When `force_keyframe` is set (first send to a
/// slot, roster change, or a host rejection), the keyframe form is emitted
/// directly.
pub fn encode(prev: &[u8], next: &[u8], dst: &mut [u8], epoch: u32, force_keyframe: bool) -> usize {
    if force_keyframe {
        return encode_keyframe(next, dst, epoch);
    }
    let n = encode_run_list(prev, next, dst, epoch);
    if n >= KEYFRAME_BYTES {
        return encode_keyframe(next, dst, epoch);
    }
    n
}

/// Apply a delta to the packed baseline `prev` in place — the inverse of
/// `encode`, used by tests to prove the round-trip (the host applies deltas in
/// production). Returns false on a structurally malformed delta
/// (degrade-to-drop, never panics). A keyframe overwrites all cells.
#[cfg(test)]
pub(crate) fn apply(prev: &mut [u8], delta: &[u8]) -> bool {
    if delta.len() < DELTA_HEADER_BYTES || prev.len() < FRAME_BYTES {
        return false;
    }
    let flags = delta[0];
    if flags & !FLAG_KEYFRAME != 0 {
        return false; // unknown flag bits
    }
    if delta[7] != GEOMETRY_ROWS || delta[8] != GEOMETRY_COLS {
        return false; // wrong geometry
    }
    let run_count = u16::from_le_bytes([delta[5], delta[6]]) as usize;
    let mut p = DELTA_HEADER_BYTES;
    let mut last_end = 0usize;
    for _ in 0..run_count {
        if p + RUN_HEADER_BYTES > delta.len() {
            return false;
        }
        let start = u16::from_le_bytes([delta[p], delta[p + 1]]) as usize;
        let run_len = u16::from_le_bytes([delta[p + 2], delta[p + 3]]) as usize;
        p += RUN_HEADER_BYTES;
        if run_len == 0 || start < last_end || start + run_len > FRAME_CELLS {
            return false; // overlap / descending / out of bounds
        }
        let n = run_len * CELL_BYTES;
        if p + n > delta.len() {
            return false; // short read
        }
        let o = start * CELL_BYTES;
        prev[o..o + n].copy_from_slice(&delta[p..p + n]);
        p += n;
        last_end = start + run_len;
    }
    p == delta.len()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::frame::{Frame, Style, CYAN, WHITE, YELLOW};

    fn packed(build: impl FnOnce(&mut Frame)) -> Vec<u8> {
        let mut f = Frame::new();
        build(&mut f);
        let mut v = vec![0u8; FRAME_BYTES];
        f.pack_into(&mut v);
        v
    }

    #[test]
    fn cell_is_24_bytes_and_frame_is_46080() {
        assert_eq!(CELL_BYTES, 24);
        assert_eq!(FRAME_BYTES, 46080);
        assert_eq!(KEYFRAME_BYTES, 46093);
    }

    #[test]
    fn no_change_is_header_only() {
        let f = packed(|_| {});
        let mut dst = vec![0u8; KEYFRAME_BYTES];
        let n = encode_run_list(&f, &f, &mut dst, 7);
        assert_eq!(n, DELTA_HEADER_BYTES);
        assert_eq!(dst[0], 0); // not keyframe
        assert_eq!(u32::from_le_bytes([dst[1], dst[2], dst[3], dst[4]]), 7); // epoch
        assert_eq!(u16::from_le_bytes([dst[5], dst[6]]), 0); // runCount
        assert_eq!((dst[7], dst[8]), (24, 80)); // geometry
    }

    #[test]
    fn keyframe_layout_and_size() {
        let next = packed(|f| f.set_rune(2, 3, 'X', Style::new(WHITE, 0)));
        let mut dst = vec![0u8; KEYFRAME_BYTES];
        let n = encode_keyframe(&next, &mut dst, 9);
        assert_eq!(n, KEYFRAME_BYTES);
        assert_eq!(dst[0], FLAG_KEYFRAME);
        assert_eq!(u32::from_le_bytes([dst[1], dst[2], dst[3], dst[4]]), 9);
        assert_eq!(u16::from_le_bytes([dst[5], dst[6]]), 1); // runCount
        assert_eq!((dst[7], dst[8]), (24, 80));
        assert_eq!(u16::from_le_bytes([dst[9], dst[10]]), 0); // start
        assert_eq!(u16::from_le_bytes([dst[11], dst[12]]), FRAME_CELLS as u16); // len
        assert_eq!(&dst[13..], &next[..]);
    }

    #[test]
    fn round_trip_apply_equals_next() {
        let prev = packed(|f| {
            f.text(0, 0, "hello", Style::new(WHITE, 0));
        });
        let next = packed(|f| {
            f.text(0, 0, "hello", Style::new(WHITE, 0));
            f.set_rune(5, 10, 'Z', Style::new(CYAN, 0));
            f.text(12, 0, "world", Style::new(YELLOW, 0));
        });
        let mut dst = vec![0u8; KEYFRAME_BYTES];
        let n = encode_run_list(&prev, &next, &mut dst, 0);
        assert!(n < KEYFRAME_BYTES);
        let mut recon = prev.clone();
        assert!(apply(&mut recon, &dst[..n]));
        assert_eq!(recon, next);
    }

    #[test]
    fn full_change_falls_back_to_keyframe() {
        let prev = packed(|f| f.fill(0, 0, 23, 79, crate::frame::Cell { rune: 'A', ..Default::default() }));
        let next = packed(|f| f.fill(0, 0, 23, 79, crate::frame::Cell { rune: 'B', ..Default::default() }));
        let mut dst = vec![0u8; KEYFRAME_BYTES];
        let n = encode(&prev, &next, &mut dst, 0, false);
        assert_eq!(n, KEYFRAME_BYTES);
        assert_eq!(dst[0], FLAG_KEYFRAME);
    }

    #[test]
    fn malformed_deltas_are_rejected_not_panicking() {
        let mut prev = packed(|_| {});
        assert!(!apply(&mut prev, &[0u8; 3])); // truncated header
        let mut d = vec![0u8; DELTA_HEADER_BYTES];
        d[0] = 0x02; // unknown flag bit
        d[7] = 24;
        d[8] = 80;
        assert!(!apply(&mut prev, &d));
        let mut d = vec![0u8; DELTA_HEADER_BYTES];
        d[7] = 25; // wrong geometry
        d[8] = 80;
        assert!(!apply(&mut prev, &d));
        let mut d = vec![0u8; DELTA_HEADER_BYTES + RUN_HEADER_BYTES];
        d[5] = 1; // runCount = 1
        d[7] = 24;
        d[8] = 80;
        d[9..11].copy_from_slice(&1919u16.to_le_bytes()); // start
        d[11..13].copy_from_slice(&5u16.to_le_bytes()); // len overruns
        assert!(!apply(&mut prev, &d));
    }
}
